frontend/lib/identification/user_code_model.dart

```diff
@@ -1,38 +1,60 @@
-import 'dart:developer';
+import 'dart:collection';
 
 import 'package:ehrenamtskarte/build_config/build_config.dart';
 import 'package:ehrenamtskarte/identification/user_code_store.dart';
 import 'package:ehrenamtskarte/proto/card.pb.dart';
 import 'package:flutter/foundation.dart';
 
+import 'package:ehrenamtskarte/sentry.dart';
+
+enum _InitializationState { uninitialized, initializing, failed, initialized }
+
 class UserCodeModel extends ChangeNotifier {
   List<DynamicUserCode> _userCodes = [];
-  bool _isInitialized = false;
+  _InitializationState _initState = _InitializationState.uninitialized;
 
   List<DynamicUserCode> get userCodes {
-    return _userCodes;
+    if (!isInitialized) {
+      throw StateError('UserCodeModel not initialized!');
+    }
+    return UnmodifiableListView(_userCodes);
   }
 
   bool get isInitialized {
-    return _isInitialized;
+    return _initState == _InitializationState.initialized;
+  }
+
+  bool get initializationFailed {
+    return _initState == _InitializationState.failed;
   }
 
   Future<void> initialize() async {
-    if (_isInitialized) {
+    if (_initState == _InitializationState.initializing || _initState == _InitializationState.initialized) {
       return;
     }
+    _initState = _InitializationState.initializing;
+    const store = UserCodeStore();
     try {
-      await UserCodeStore().importLegacyCard();
-      _userCodes = await const UserCodeStore().load();
-    } on Exception catch (e) {
-      log('Failed to initialize activation code from secure storage.', error: e);
+      DynamicUserCode? legacyCode = await store.loadAndDeleteLegacyCard();
+      var userCodes = await store.load();
+      if (legacyCode != null && !isAlreadyInList(userCodes, legacyCode.info)) {
+        userCodes.add(legacyCode);
+        await store.store(userCodes);
+      }
+      _userCodes = userCodes;
+      _initState = _InitializationState.initialized;
+    } catch (e, stackTrace) {
+      reportError('Failed to initialize activation code from secure storage: $e', stackTrace);
+      _initState = _InitializationState.failed;
     } finally {
-      _isInitialized = true;
       notifyListeners();
     }
   }
 
   void insertCode(DynamicUserCode code) {
+    if (!isInitialized) {
+      throw StateError('UserCodeModel not initialized!');
```

Comment:
Maybe you could move this error text to a constant that is reused, to avoid duplication

Reply:
Done